src/Sarif/Writers/ResultLogJsonWriter.cs

```diff
@@ -124,38 +161,50 @@ public void WriteResult(Result result)
                 throw new ArgumentNullException("result");
             }
 
-            this.EnsureNotDisposed();
-            if (_writeState == State.Initial)
+            EnsureStateNotAlreadySet(Conditions.Disposed | Conditions.ResultsClosed);
+
+            if ((_writeConditions & Conditions.ResultsInitialized) != Conditions.ResultsInitialized)
             {
-                throw new InvalidOperationException(SarifResources.CannotWriteResultToolInfoMissing);
+                OpenResults();
             }
 
-            Debug.Assert(_writeState == State.WritingResults);
             _serializer.Serialize(_jsonWriter, result, typeof(Result));
         }
 
         public void CloseResults()
         {
+            EnsureStateNotAlreadySet(Conditions.Disposed);
```

Comment:
No, we allow some resilience for writers that stream individual results without trying to manage the results array. I added a comment.